api/client/service/update.go

```diff
@@ -358,12 +358,23 @@ func updateMounts(flags *pflag.FlagSet, mounts *[]swarm.Mount) {
 }
 
 func updatePorts(flags *pflag.FlagSet, portConfig *[]swarm.PortConfig) {
+	portSet := make(map[swarm.PortConfig]struct{})
+	for _, entry := range *portConfig {
+		portSet[entry] = struct{}{}
+	}
+
 	if flags.Changed(flagPublishAdd) {
 		values := flags.Lookup(flagPublishAdd).Value.(*opts.ListOpts).GetAll()
 		ports, portBindings, _ := nat.ParsePortSpecs(values)
 
 		for port := range ports {
-			*portConfig = append(*portConfig, convertPortToPortConfig(port, portBindings)...)
+			newConfigs := convertPortToPortConfig(port, portBindings)
```

Comment:
Make sure to sort this, as well, or random orderings could cause unnecessary updates.